cdap-app-templates/cdap-etl/cdap-data-pipeline/src/test/java/co/cask/cdap/datapipeline/LineageOperationProcessorTest.java

```diff
@@ -430,4 +497,247 @@ public void testComplexMerge() {
+                                                     "body"));
+    stageOperations.put("n1", pipelineOperations);
+    pipelineOperations = new ArrayList<>();
+
+    // Output of the following operation is redundant since it will be overwritten by the parse operation
+    pipelineOperations.add(new PipelineTransformOperation("redundant_parse", "parsing data",
+                                                          Collections.singletonList("body"), "name"));
+
+    pipelineOperations.add(new PipelineTransformOperation("parse", "parsing data", Collections.singletonList("body"),
+                                                          Arrays.asList("name", "address", "zip")));
+    stageOperations.put("n2", pipelineOperations);
+    pipelineOperations = new ArrayList<>();
+    pipelineOperations.add(new PipelineWriteOperation("write", "writing data", EndPoint.of("default", "file2"),
+                                                      "name", "address", "zip"));
+    stageOperations.put("n3", pipelineOperations);
+
+    Map<String, StageInputOutput> stageInputOutputs = new HashMap<>();
+    stageInputOutputs.put("n1", new StageInputOutput(Collections.emptySet(),
+                                                     new HashSet<>(Arrays.asList("offset", "body"))));
+    stageInputOutputs.put("n2", new StageInputOutput(new HashSet<>(Arrays.asList("offset", "body")),
+                                                     new HashSet<>(Arrays.asList("name", "address", "zip"))));
+    stageInputOutputs.put("n3", new StageInputOutput(new HashSet<>(Arrays.asList("name", "address", "zip")),
+                                                     Collections.emptySet()));
+
+    try {
+      LineageOperationsProcessor processor = new LineageOperationsProcessor(connections, stageInputOutputs,
+                                                                            stageOperations, Collections.emptySet());
+      Assert.fail();
+    } catch (InvalidLineageException e) {
+      // Exception message:
+      // "Outputs of following operations are neither used by subsequent operations in " +
+      //  "that stage nor are part of the output schema of that stage: " +
+      //  "<stage:n2, [operation:redundant_parse, field:name]>. ";
+
+      Map<String, InvalidFieldOperations> expected = new HashMap<>();
+      InvalidFieldOperations invalids =
+        new InvalidFieldOperations(Collections.emptyMap(),
+                                   ImmutableMap.of("name", Collections.singletonList("redundant_parse")));
+
+      expected.put("n2", invalids);
+
+      Map<String, InvalidFieldOperations> invalidFieldOperations = e.getInvalidFieldOperations();
+      Assert.assertEquals(expected, invalidFieldOperations);
+    }
+  }
+
+  @Test
+  public void testRedundantOutputUsedAsInput() {
+    // n1-->n2-->n3
+    Set<Connection> connections = new HashSet<>();
+    connections.add(new Connection("n1", "n2"));
+    connections.add(new Connection("n2", "n3"));
```

Comment:
you can move all these tests to a LineageOperationsProcessorTest, where you don't need to create these irrelevant data structures.

Reply:
fixed.